File: examples/scripts/python/activity_recording.py
#!/usr/bin/env python3
"""Model Health Python SDK — capture workflow demo.

Walks through the full capture workflow:
  1. Create a session (QR code saved locally for pairing)
  2. Select an existing subject or create a new one
  3. Calibrate cameras using a checkerboard pattern
  4. Calibrate the subject (neutral standing pose)
  5. Record one or more movement activities in a loop
  6. For each activity: wait for upload/processing, optionally run analysis,
     then optionally update activity metadata
  7. After each activity, choose to record another or quit

Requires cameras to be connected and ready via the Model Health companion iOS app.

Usage:
    activity_recording.py [<api_key>]
"""

import os
import sys
import time
import urllib.request

from docopt import docopt

from modelhealth import (
    ActivityConfig,
    ActivityStatus,
    ActivityStatusAnalyzing,
    ActivityStatusUploading,
    ActivityType,
    AnalysisDataType,
    AnalysisStatus,
    CalibrationStatus,
    CalibrationStatusProcessing,
    CalibrationStatusUploading,
    CheckerboardDetails,
    CheckerboardPlacement,
    FilterFrequencyHz,
    ModelHealthError,
    ModelHealthClient,
    RecordingConfig,
    SessionFramerate,
    SubjectParameters,
)
from _prompts import confirm, pick_one
from _utils import ANALYSIS_DATA_EXT, DOWNLOADS_DIR, load_api_key, poll_analysis, save_file

# ---------------------------------------------------------------------------
# Activity type options for automatic analysis after recording
# ---------------------------------------------------------------------------

_ACTIVITY_TYPES = [
    (None,                               "None — skip automatic analysis"),
    (ActivityType.counter_movement_jump, "Counter Movement Jump"),
    (ActivityType.gait,                  "Overground Walking"),
    (ActivityType.treadmill_gait,        "Treadmill Walking"),
    (ActivityType.treadmill_running,     "Treadmill Running"),
    (ActivityType.overground_running,    "Overground Running"),
    (ActivityType.sit_to_stand,          "Sit-to-Stand Transfer"),
    (ActivityType.squats,                "Squats"),
    (ActivityType.range_of_motion,       "Range of Motion"),
    (ActivityType.drop_jump,             "Drop Vertical Jump"),
    (ActivityType.hop,                   "Hop Test"),
    (ActivityType.change_of_direction,   "5-0-5 Test"),
    (ActivityType.cut,                   "Cutting Maneuver"),
    (ActivityType.sprint,                "Sprint"),
    (ActivityType.lateral_stepdown,      "Lateral Step Down"),
    (ActivityType.lunge,                 "Lunge"),
]

# ---------------------------------------------------------------------------
# Per-recording config options
# ---------------------------------------------------------------------------

_FRAMERATES = [
    (None,                     "Default"),
    (SessionFramerate.fps_60,  "60 fps"),
    (SessionFramerate.fps_120, "120 fps"),
    (SessionFramerate.fps_240, "240 fps"),
]

_FILTER_FREQUENCIES = [
    (None,                   "Default"),
    (FilterFrequencyHz(6),   "6 Hz"),
    (FilterFrequencyHz(10),  "10 Hz"),
    (FilterFrequencyHz(20),  "20 Hz"),
]

# ---------------------------------------------------------------------------
# Checkerboard presets
# ---------------------------------------------------------------------------

_CHECKERBOARD_PRESETS = [
    (4, 5, 35, "4 × 5  —  35 mm squares  (standard A4)"),
    (4, 5, 50, "4 × 5  —  50 mm squares  (large A3)"),
    (None, None, None, "Other — enter manually"),
]

# ---------------------------------------------------------------------------
# Calibration callback
# ---------------------------------------------------------------------------

_CB_WIDTH = 48


def _calibration_callback(status):
    if isinstance(status, CalibrationStatusUploading):
        msg = f"  Uploading ({status.uploaded}/{status.total} cameras)..."
        print(msg.ljust(_CB_WIDTH), end="\r", flush=True)
    elif isinstance(status, CalibrationStatusProcessing):
        pct = f"{status.percent}%" if status.percent is not None else "--%"
        msg = f"  Processing ({pct})..."
        print(msg.ljust(_CB_WIDTH), end="\r", flush=True)
    elif status == CalibrationStatus.recording:
        print("  Recording...".ljust(_CB_WIDTH), end="\r", flush=True)
    elif status == CalibrationStatus.done:
        print("  Done.".ljust(_CB_WIDTH))


# ---------------------------------------------------------------------------
# Activity polling helper
# ---------------------------------------------------------------------------

def _poll_activity(client, activity, interval=5):
    """Block until the activity finishes uploading and processing."""
    while True:
        status = client.activity_status(activity)
        if isinstance(status, ActivityStatusUploading):
            print(
                f"  Uploading ({status.uploaded}/{status.total} cameras)...  ",
                end="\r", flush=True,
            )
        elif status == ActivityStatus.processing:
            print("  Processing...                              ", end="\r", flush=True)
        else:
            print()
            return status
        time.sleep(interval)


# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------

def _connect(api_key):
    print("Connecting...")
    try:
        return ModelHealthClient(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")


# ---------------------------------------------------------------------------
# Session / camera pairing
# ---------------------------------------------------------------------------

def _create_session_and_save_qr_code(client):
    print("\nCreating session...")
    try:
        session = client.create_session()
    except ModelHealthError as exc:
        sys.exit(f"Failed to create session: {exc}")
    print(f"  Session ID: {session.id}")

    if not session.qrcode:
        sys.exit("Session has no QR code — cannot pair cameras.")

    os.makedirs(DOWNLOADS_DIR, exist_ok=True)
    qr_path = os.path.join(DOWNLOADS_DIR, "qr-code.png")
    try:
        urllib.request.urlretrieve(session.qrcode, qr_path)
        print(f"  QR code saved to: {qr_path}")
    except Exception as exc:
        sys.exit(f"Failed to download QR code: {exc}")

    return session


def _wait_for_camera_pairing():
    print("  Pair your cameras using the Model Health companion iOS app before continuing.")
    input("\nPress Enter when cameras are ready...")


# ---------------------------------------------------------------------------
# Camera calibration
# ---------------------------------------------------------------------------

def _configure_checkerboard():
    print("\nCheckerboard configuration:\n")
    preset = pick_one(
        _CHECKERBOARD_PRESETS,
        "Select checkerboard",
        lambda p: p[3],
    )
    rows, columns, square_size, _ = preset

    if rows is None:
        rows        = int(input("  Internal rows: ").strip())
        columns     = int(input("  Internal columns: ").strip())
        square_size = int(input("  Square size (mm): ").strip())

    print("\nCheckerboard placement:\n")
    placement_value, _ = pick_one(
        [
            (CheckerboardPlacement.perpendicular, "Perpendicular (upright, facing cameras)"),
            (CheckerboardPlacement.parallel,      "Parallel (flat on the floor)"),
        ],
        "Select placement",
        lambda p: p[1],
    )

    return CheckerboardDetails(
        rows=rows,
        columns=columns,
        square_size=square_size,
        placement=placement_value,
    )


def _calibrate_cameras(client, session, checkerboard):
    input("\nPress Enter to start camera calibration...")
    print("Calibrating cameras...")
    try:
        client.calibrate_camera(session, checkerboard, _calibration_callback)
    except ModelHealthError as exc:
        sys.exit(f"Camera calibration failed: {exc}")
    print("Camera calibration complete.")


# ---------------------------------------------------------------------------
# Subject
# ---------------------------------------------------------------------------

def _pick_or_create_subject(client):
    print("\nFetching subjects...")
    try:
        subjects = client.subject_list()
    except ModelHealthError as exc:
        sys.exit(f"Failed to fetch subjects: {exc}")

    if subjects and confirm(f"Found {len(subjects)} subject(s). Select an existing one?", default=True):
        print()
        subject = pick_one(
            subjects,
            "Select subject",
            lambda s: f"{s.name}  (ID {s.id})",
        )
        print(f"  Using: {subject.name}")
        return subject

    print("\nNew subject details:")
    name   = input("  Name: ").strip() or "Anonymous"
    weight = float(input("  Weight (kg): ").strip())
    height = float(input("  Height (cm): ").strip())

    params = SubjectParameters(name=name, weight=weight, height=height)
    print("Creating subject...")
    try:
        subject = client.create_subject(params)
    except ModelHealthError as exc:
        sys.exit(f"Failed to create subject: {exc}")
    print(f"  Subject created: {subject.name} (ID {subject.id})")
    return subject


def _calibrate_subject(client, subject, session):
    input(f"\nAsk {subject.name} to stand in the neutral pose, then press Enter...")
    print("Calibrating subject...")
    try:
        client.calibrate_subject(subject, session, _calibration_callback)
    except ModelHealthError as exc:
        sys.exit(f"Subject calibration failed: {exc}")
    print("Subject calibration complete.")


# ---------------------------------------------------------------------------
# Single recording loop iteration
# ---------------------------------------------------------------------------

def _prompt_recording_config():
    """Returns (activity_name, activity_type_value, activity_type_label, recording_config)."""
    activity_name = input("\nActivity name (e.g. cmj, squat): ").strip() or "activity"

    print("\nAutomatic analysis (optional):\n")
    activity_type_value, activity_type_label = pick_one(
        _ACTIVITY_TYPES,
        "Select activity type",
        lambda t: t[1],
    )

    print("\nFramerate override (optional):\n")
    framerate_value, _ = pick_one(_FRAMERATES, "Select framerate", lambda t: t[1])

    print("\nFilter frequency override (optional):\n")
    filter_value, _ = pick_one(_FILTER_FREQUENCIES, "Select filter frequency", lambda t: t[1])

    recording_config = None
    if framerate_value is not None or filter_value is not None:
        recording_config = RecordingConfig(framerate=framerate_value, filter_frequency=filter_value)

    return activity_name, activity_type_value, activity_type_label, recording_config


def _start_recording(client, session, subject, activity_name, activity_type_value, recording_config):
    input(f"\nAsk {subject.name} to get ready, then press Enter to start recording...")
    print("Recording...")
    try:
        activity = client.start_recording(
            activity_name,
            session,
            ActivityConfig(activity_type=activity_type_value, config=recording_config)
        )
    except ModelHealthError as exc:
        print(f"Failed to start recording: {exc}")
        return None

    print(f"  Recording started (activity {activity.id}).")
    return activity


def _stop_recording(client, session):
    input("\nPress Enter when the movement is complete to stop recording...")
    print("Stopping recording...")
    try:
        client.stop_recording(session)
    except ModelHealthError as exc:
        print(f"Failed to stop recording: {exc}")
        return False

    print("Recording stopped. Videos are uploading.")
    return True


def _wait_and_process_results(client, activity, activity_type_label):
    """Waits for upload/processing and, if automatic analysis was requested,
    waits for it to complete and downloads the report.

    Always returns the activity (fresh, if analysis ran and completed).
    """
    print("\nWaiting for upload and processing...")
    status = _poll_activity(client, activity)

    if isinstance(status, ActivityStatusAnalyzing):
        return _wait_for_analysis_and_download_report(client, activity, status.task, activity_type_label)
    elif status == ActivityStatus.ready:
        print(f"Activity is ready. ID: {activity.id}")
        print("Run activity_analysis.py to analyze this activity.")
        return activity
    else:
        print(f"Activity did not reach ready state (status: {status}).")
        return activity


def _wait_for_analysis_and_download_report(client, activity, task, activity_type_label):
    print(f"Activity is ready. Automatic '{activity_type_label}' analysis has started.")
    print("\nWaiting for analysis to complete...")
    result_status = poll_analysis(client, task)
    if result_status != AnalysisStatus.completed:
        print(f"Analysis did not complete (status: {result_status}).")
        return activity

    print("Analysis complete.")
    activity = client.fetch_activity(activity.id)
    results = client.analysis_data_for_activity(activity, [AnalysisDataType.report])
    slug = (activity.name or activity.id).replace(" ", "_")
    print("\nDownloading report...")
    for r in results:
        ext = ANALYSIS_DATA_EXT.get(r.type, "bin")
        path = save_file(f"{slug}_{r.type}.{ext}", r.data)
        print(f"  Saved {path}")

    return activity


def _update_activity_metadata(client, activity):
    # Re-fetch first: analysis auto-generates tags server-side, and update_activity
    # merges add/remove_tags on top of the local activity's tags. Without a fresh
    # fetch, the merge starts from a stale tag set and wipes the auto-generated tags.
    try:
        activity = client.fetch_activity(activity.id)
    except ModelHealthError as exc:
        print(f"Failed to refresh activity: {exc}")

    print("\nUpdate activity (optional):")
    current_tags = ", ".join(activity.tags) if activity.tags else "(none)"
    print(f"  Current tags: {current_tags}")

    new_name = input(f"  New name (press Enter to keep '{activity.name or activity.id}'): ").strip() or None

    add_input = input("  Tags to add, comma-separated (press Enter to skip): ").strip()
    add_tags = [t.strip() for t in add_input.split(",") if t.strip()] if add_input else []

    remove_input = input("  Tags to remove, comma-separated (press Enter to skip): ").strip()
    remove_tags = [t.strip() for t in remove_input.split(",") if t.strip()] if remove_input else []

    if new_name or add_tags or remove_tags:
        print("Updating activity...")
        try:
            activity = client.update_activity(
                activity,
                ActivityConfig(name=new_name, add_tags=add_tags, remove_tags=remove_tags),
            )
        except ModelHealthError as exc:
            print(f"Failed to update activity: {exc}")
        else:
            print(f"  Updated: {activity.name or activity.id}")

    return activity


def _record_one(client, session, subject):
    """Run a single record → process → (optionally analyse) → (optionally update) cycle.

    Returns True to continue recording, False to quit.
    """
    activity_name, activity_type_value, activity_type_label, recording_config = _prompt_recording_config()

    activity = _start_recording(client, session, subject, activity_name, activity_type_value, recording_config)
    if activity is None:
        return confirm("\nRecord another activity?", default=True)

    if not _stop_recording(client, session):
        return confirm("\nRecord another activity?", default=True)

    activity = _wait_and_process_results(client, activity, activity_type_label)
    _update_activity_metadata(client, activity)

    return confirm("\nRecord another activity?", default=True)


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    client = _connect(api_key)

    session = _create_session_and_save_qr_code(client)
    _wait_for_camera_pairing()

    checkerboard = _configure_checkerboard()
    _calibrate_cameras(client, session, checkerboard)

    while True:
        subject = _pick_or_create_subject(client)
        _calibrate_subject(client, subject, session)

        # Recording loop
        while _record_one(client, session, subject):
            pass

        if not confirm("\nCalibrate another subject with the same camera setup?", default=True):
            break
        session = client.new_session_from_session(session)

    print("\nDone.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))

File: examples/scripts/ts/activity_recording.ts
/**
 * Model Health TypeScript SDK — capture workflow.
 * Mirrors examples/python/activity_recording.py.
 *
 * Usage:
 *   npx tsx activity_recording.ts [<api_key>]
 */

import { ModelHealthClient, ActivityType } from '@modelhealth/modelhealth';
import type {
  CheckerboardDetails, CheckerboardPlacement,
  CalibrationStatus, ActivityStatus, Analysis, RecordingConfig,
} from '@modelhealth/modelhealth';
import {
  loadApiKey, ANALYSIS_DATA_EXT,
  pickOne, confirm, prompt, saveFile, pollAnalysis, sleep, closePrompts,
} from './_shared.js';

const CB_WIDTH = 48;

function calibrationCallback(status: CalibrationStatus): void {
  let msg: string;
  switch (status.type) {
    case 'recording':
      msg = '  Recording...';
      break;
    case 'uploading':
      msg = `  Uploading (${status.uploaded}/${status.total} cameras)...`;
      break;
    case 'processing': {
      const pct = status.percent != null ? `${status.percent}%` : '--%';
      msg = `  Processing (${pct})...`;
      break;
    }
    case 'done':
      process.stdout.write('  Done.'.padEnd(CB_WIDTH) + '\n');
      return;
  }
  process.stdout.write(msg.padEnd(CB_WIDTH) + '\r');
}

interface CheckerboardPreset {
  rows: number | null;
  columns: number | null;
  squareSize: number | null;
  label: string;
}

const CHECKERBOARD_PRESETS: CheckerboardPreset[] = [
  { rows: 4, columns: 5, squareSize: 35, label: '4 × 5  —  35 mm squares  (standard A4)' },
  { rows: 4, columns: 5, squareSize: 50, label: '4 × 5  —  50 mm squares  (large A3)' },
  { rows: null, columns: null, squareSize: null, label: 'Other — enter manually' },
];

interface ActivityTypeOption {
  type: ActivityType | null;
  label: string;
}

const ACTIVITY_TYPE_OPTIONS: ActivityTypeOption[] = [
  { type: null,                               label: 'None — skip automatic analysis'  },
  { type: ActivityType.CounterMovementJump,   label: 'Counter Movement Jump'           },
  { type: ActivityType.Gait,                  label: 'Overground Walking'              },
  { type: ActivityType.TreadmillGait,         label: 'Treadmill Walking'               },
  { type: ActivityType.TreadmillRunning,      label: 'Treadmill Running'               },
  { type: ActivityType.OvergroundRunning,     label: 'Overground Running'              },
  { type: ActivityType.SitToStand,            label: 'Sit-to-Stand Transfer'           },
  { type: ActivityType.Squats,                label: 'Squats'                          },
  { type: ActivityType.RangeOfMotion,         label: 'Range of Motion'                 },
  { type: ActivityType.DropJump,              label: 'Drop Vertical Jump'              },
  { type: ActivityType.Hop,                   label: 'Hop Test'                        },
  { type: ActivityType.ChangeOfDirection,     label: '5-0-5 Test'                      },
  { type: ActivityType.Cut,                   label: 'Cutting Maneuver'                },
  { type: ActivityType.Sprint,                label: 'Sprint'                          },
  { type: ActivityType.LateralStepdown,       label: 'Lateral Step Down'               },
  { type: ActivityType.Lunge,                 label: 'Lunge'                           },
];

interface FramerateOption {
  value: number | null;
  label: string;
}

const FRAMERATE_OPTIONS: FramerateOption[] = [
  { value: null, label: 'Default'  },
  { value: 60,   label: '60 fps'  },
  { value: 120,  label: '120 fps' },
  { value: 240,  label: '240 fps' },
];

interface FilterFrequencyOption {
  value: { type: 'hz'; value: number } | null;
  label: string;
}

const FILTER_FREQUENCY_OPTIONS: FilterFrequencyOption[] = [
  { value: null,                       label: 'Default' },
  { value: { type: 'hz', value: 6 },   label: '6 Hz'    },
  { value: { type: 'hz', value: 10 },  label: '10 Hz'   },
  { value: { type: 'hz', value: 20 },  label: '20 Hz'   },
];

async function pollActivity(
  client: ModelHealthClient,
  activity: Parameters<typeof client.activityStatus>[0]
): Promise<ActivityStatus> {
  while (true) {
    const status = await client.activityStatus(activity);
    if (status.type === 'uploading') {
      process.stdout.write(`  Uploading (${status.uploaded}/${status.total} cameras)...  \r`);
    } else if (status.type === 'processing') {
      process.stdout.write('  Processing...                              \r');
    } else {
      process.stdout.write('\n');
      return status;
    }
    await sleep(5_000);
  }
}

async function connect(apiKey: string): Promise<ModelHealthClient> {
  console.log('Connecting...');
  const client = new ModelHealthClient({ apiKey, autoInit: false });
  await client.init();
  return client;
}

async function createSessionAndSaveQrCode(client: ModelHealthClient) {
  console.log('\nCreating session...');
  const session = await client.createSession();
  console.log(`  Session ID: ${session.id}`);

  if (!session.qrcode) {
    console.error('Session has no QR code — cannot pair cameras.');
    process.exit(1);
  }

  const qrRes = await fetch(session.qrcode);
  const qrData = new Uint8Array(await qrRes.arrayBuffer());
  const qrPath = saveFile('qr-code.png', qrData);
  console.log(`  QR code saved to: ${qrPath}`);

  return session;
}

async function waitForCameraPairing() {
  console.log('  Pair your cameras using the Model Health companion iOS app before continuing.');
  await prompt('\nPress Enter when cameras are ready...');
}

async function configureCheckerboard(): Promise<CheckerboardDetails> {
  console.log('\nCheckerboard configuration:\n');
  const preset = await pickOne(CHECKERBOARD_PRESETS, 'Select checkerboard', p => p.label);

  let rows: number;
  let columns: number;
  let squareSize: number;

  if (preset.rows !== null && preset.columns !== null && preset.squareSize !== null) {
    rows = preset.rows;
    columns = preset.columns;
    squareSize = preset.squareSize;
  } else {
    rows      = parseInt((await prompt('  Internal rows: ')).trim(), 10);
    columns   = parseInt((await prompt('  Internal columns: ')).trim(), 10);
    squareSize = parseInt((await prompt('  Square size (mm): ')).trim(), 10);
  }

  console.log('\nCheckerboard placement:\n');
  const placement = await pickOne<CheckerboardPlacement>(
    ['perpendicular', 'parallel'],
    'Select placement',
    p => p === 'perpendicular' ? 'Perpendicular (upright, facing cameras)' : 'Parallel (flat on the floor)'
  );

  return { rows, columns, squareSize, placement };
}

async function calibrateCameras(
  client: ModelHealthClient,
  session: Awaited<ReturnType<typeof createSessionAndSaveQrCode>>,
  checkerboard: CheckerboardDetails
) {
  await prompt('\nPress Enter to start camera calibration...');
  console.log('Calibrating cameras...');
  await client.calibrateCamera(session, checkerboard, calibrationCallback);
  console.log('Camera calibration complete.');
}

async function pickOrCreateSubject(client: ModelHealthClient) {
  console.log('\nFetching subjects...');
  const subjects = await client.subjectList();

  if (subjects.length > 0 && await confirm(`Found ${subjects.length} subject(s). Select an existing one?`, true)) {
    console.log();
    const subject = await pickOne(subjects, 'Select subject', s => `${s.name}  (ID ${s.id})`);
    console.log(`  Using: ${subject.name}`);
    return subject;
  }

  console.log('\nNew subject details:');
  const name   = (await prompt('  Name: ')).trim() || 'Anonymous';
  const weight = parseFloat((await prompt('  Weight (kg): ')).trim());
  const height = parseFloat((await prompt('  Height (cm): ')).trim());
  console.log('Creating subject...');
  const subject = await client.createSubject({ name, weight, height });
  console.log(`  Subject created: ${subject.name} (ID ${subject.id})`);
  return subject;
}

async function calibrateSubject(
  client: ModelHealthClient,
  subject: Awaited<ReturnType<typeof pickOrCreateSubject>>,
  session: Awaited<ReturnType<typeof createSessionAndSaveQrCode>>
) {
  await prompt(`\nAsk ${subject.name} to stand in the neutral pose, then press Enter...`);
  console.log('Calibrating subject...');
  await client.calibrateSubject(subject, session, calibrationCallback);
  console.log('Subject calibration complete.');
}

async function main() {
  const args = process.argv.slice(2);
  const client = await connect(loadApiKey(args[0]));

  let session = await createSessionAndSaveQrCode(client);
  await waitForCameraPairing();

  const checkerboard = await configureCheckerboard();
  await calibrateCameras(client, session, checkerboard);

  for (;;) {
    const subject = await pickOrCreateSubject(client);
    await calibrateSubject(client, subject, session);

    // Recording loop
    do {
      await recordOne(client, session, subject);
    } while (await confirm('\nRecord another activity?', true));

    if (!(await confirm('\nCalibrate another subject with the same camera setup?', true))) {
      break;
    }
    session = await client.newSessionFromSession(session);
  }

  console.log('\nDone.');
}

async function promptRecordingSetup() {
  const activityName = (await prompt('\nActivity name (e.g. cmj, squat): ')).trim() || 'activity';

  console.log('\nAutomatic analysis (optional):\n');
  const selectedType = await pickOne(ACTIVITY_TYPE_OPTIONS, 'Select activity type', t => t.label);

  console.log('\nFramerate override (optional):\n');
  const selectedFramerate = await pickOne(FRAMERATE_OPTIONS, 'Select framerate', o => o.label);

  console.log('\nFilter frequency override (optional):\n');
  const selectedFilter = await pickOne(FILTER_FREQUENCY_OPTIONS, 'Select filter frequency', o => o.label);

  let recordingConfig: RecordingConfig | undefined;
  if (selectedFramerate.value !== null || selectedFilter.value !== null) {
    recordingConfig = {};
    if (selectedFramerate.value !== null) recordingConfig.framerate = selectedFramerate.value as any;
    if (selectedFilter.value !== null) recordingConfig.filterFrequency = selectedFilter.value;
  }
  const config = (selectedType.type || recordingConfig)
    ? { activityType: selectedType.type ?? undefined, config: recordingConfig }
    : undefined;

  return { activityName, selectedType, config };
}

async function startRecording(
  client: ModelHealthClient,
  session: Parameters<typeof client.startRecording>[1],
  subject: { name: string },
  activityName: string,
  config: Parameters<typeof client.startRecording>[2]
) {
  await prompt(`\nAsk ${subject.name} to get ready, then press Enter to start recording...`);
  console.log('Recording...');
  try {
    const activity = await client.startRecording(activityName, session, config);
    console.log(`  Recording started (activity ${activity.id}).`);
    return activity;
  } catch (err: any) {
    console.error(`Failed to start recording: ${err.message ?? err}`);
    return undefined;
  }
}

async function stopRecording(client: ModelHealthClient, session: Parameters<typeof client.stopRecording>[0]) {
  await prompt('\nPress Enter when the movement is complete to stop recording...');
  console.log('Stopping recording...');
  try {
    await client.stopRecording(session);
  } catch (err: any) {
    console.error(`Failed to stop recording: ${err.message ?? err}`);
    return false;
  }
  console.log('Recording stopped. Videos are uploading.');
  return true;
}

/**
 * Waits for upload/processing and, if automatic analysis was requested,
 * waits for it to complete and downloads the report. Always returns the
 * activity (fresh, if analysis ran and completed).
 */
async function waitAndProcessResults(
  client: ModelHealthClient,
  activity: Awaited<ReturnType<typeof client.startRecording>>,
  activityTypeLabel: string
) {
  console.log('\nWaiting for upload and processing...');
  const finalStatus = await pollActivity(client, activity);

  if (finalStatus.type === 'analyzing') {
    const task: Analysis = { taskId: finalStatus.taskId };
    return waitForAnalysisAndDownloadReport(client, activity, task, activityTypeLabel);
  } else if (finalStatus.type === 'ready') {
    console.log(`Activity is ready. ID: ${activity.id}`);
    console.log('Run activity_analysis.ts to analyze this activity.');
    return activity;
  } else {
    console.log(`Activity did not reach ready state (status: ${finalStatus.type}).`);
    return activity;
  }
}

async function waitForAnalysisAndDownloadReport(
  client: ModelHealthClient,
  activity: Awaited<ReturnType<typeof client.startRecording>>,
  task: Analysis,
  activityTypeLabel: string
) {
  console.log(`Activity is ready. Automatic '${activityTypeLabel}' analysis has started.`);
  console.log('\nWaiting for analysis to complete...');
  const resultStatus = await pollAnalysis(client, task);

  if (resultStatus.type !== 'completed') {
    console.log(`Analysis did not complete (status: ${resultStatus.type}).`);
    return activity;
  }

  console.log('Analysis complete.');
  const freshActivity = await client.fetchActivity(activity.id);
  const results = await client.analysisDataForActivity(freshActivity, ['report']);
  const slug = (freshActivity.name ?? freshActivity.id).replace(/ /g, '_');

  console.log('\nDownloading report...');
  for (const r of results) {
    const ext = ANALYSIS_DATA_EXT[r.type] ?? 'bin';
    const p = saveFile(`${slug}_${r.type}.${ext}`, r.data);
    console.log(`  Saved ${p}`);
  }

  return freshActivity;
}

// Re-fetch first: analysis auto-generates tags server-side, and updateActivity
// merges add/remove tags on top of the local activity's tags. Without a fresh
// fetch, the merge starts from a stale tag set and wipes the auto-generated tags.
async function updateActivityMetadata(
  client: ModelHealthClient,
  activity: Awaited<ReturnType<typeof waitAndProcessResults>>
) {
  let currentActivity = activity;
  try {
    currentActivity = await client.fetchActivity(activity.id);
  } catch (err: any) {
    console.error(`Failed to refresh activity: ${err.message ?? err}`);
  }

  console.log('\nUpdate activity (optional):');
  const currentTags = currentActivity.tags?.length ? currentActivity.tags.join(', ') : '(none)';
  console.log(`  Current tags: ${currentTags}`);
  const newName = (await prompt(`  New name (press Enter to keep '${currentActivity.name ?? currentActivity.id}'): `)).trim() || undefined;
  const addInput = (await prompt('  Tags to add, comma-separated (press Enter to skip): ')).trim();
  const addTags = addInput ? addInput.split(',').map(t => t.trim()).filter(Boolean) : [];
  const removeInput = (await prompt('  Tags to remove, comma-separated (press Enter to skip): ')).trim();
  const removeTags = removeInput ? removeInput.split(',').map(t => t.trim()).filter(Boolean) : [];
  if (newName || addTags.length || removeTags.length) {
    console.log('Updating activity...');
    try {
      const updated = await client.updateActivity(currentActivity, { name: newName, addTags, removeTags });
      console.log(`  Updated: ${updated.name ?? updated.id}`);
    } catch (err: any) {
      console.error(`Failed to update activity: ${err.message ?? err}`);
    }
  }
}

async function recordOne(
  client: ModelHealthClient,
  session: Parameters<typeof client.startRecording>[1],
  subject: { name: string }
): Promise<void> {
  const { activityName, selectedType, config } = await promptRecordingSetup();

  const activity = await startRecording(client, session, subject, activityName, config);
  if (!activity) return;

  if (!(await stopRecording(client, session))) return;

  const currentActivity = await waitAndProcessResults(client, activity, selectedType.label);
  await updateActivityMetadata(client, currentActivity);
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);
